Implement Mythic+ support

tdaugaard opened the pull request:
As discussed in issue #43, here's a preliminary implementation of Mythic+/Keystone support with time skips for trash/boss segments and what have we.

It's a draft for now, but I'm opening it to talk about the implementation/features/etc.

tdaugaard commented:
Be aware that I have no previous experience with React, so I might've done something in a way that isn't great - I'm not sure. I've googled things I wasn't sure of and such, so it should look somewhat okay.

aza547 commented:
Thanks, I'll try review this today. Had a quick skim this morning. Thomas, you should DM be on discord and join our development channel if you plan to stay involved. Also good place to discuss things like this!

tdaugaard commented:
If you want me to split it up into several PRs, or if there are parts you don't want/like/have comments about, feel free to let me know - I'll sort it out however you'd like it :)

tdaugaard commented:
I've tested it pretty thoroughly with M+ dungeons and believe I've found and fixed all bugs related to that.

I've split some commits into more which might ease the review process.

I have NOT tested this in raid or PvP content and therefore don't know if any of the changes has created bugs for those, but I _don't_ think so. I've tried to keep the changes compatible with the current code.

aza547 commented:
Having a look at this now - initial observation is there is no tests, can you check in an example dungeon log and add a dungeon test? See all_tests.py for existing infrastructure.

aza547 commented on src/renderer/Layout.tsx:
```diff
@@ -164,23 +186,6 @@ export default function Layout() {
     )
   };
 
-  /**
```
Awesome to get rid of this!

aza547 commented on src/renderer/VideoButton.tsx:
```diff
@@ -79,7 +88,7 @@ export default function VideoButton(props: any) {
   /**
   * Delete a video.
   */
-  function deleteVideo(filePath: string) {6
```
as if I checked this into main... :| 
this is why I need to write some unit tests

tdaugaard replied:
I actually thought that _I_ had made a mistake and couldn't for the life of me figure out where, but accepted that it just needed to be fixed regardless. 🤣

aza547 replied:
hilariously it somehow isn't broken... no idea why

tdaugaard replied:
Just javascript things, really. Javascript will take what you write and do the best it can, mostly ignoring things that doesn't make sense. 🤷‍♂️🤣

aza547 commented on src/renderer/rendererutils.ts:
```diff
@@ -26,24 +7,16 @@
-    }
+    switch (category) {
+        case VideoCategory.MythicPlus:
+            return isGoodResult ? "Timed" : "Fail";
```
Deplete better than fail?

tdaugaard replied:
Yes, that's what I was looking for! Thanks.

tdaugaard replied:
Fixed in 8a2d3cf

aza547 commented on src/renderer/rendererutils.ts:
```diff
@@ -26,24 +7,16 @@
 
     // Not sure how we can decide who won or lost yet. 
     // Combat log doesn't make it obvious.
-    if ((category == "Battlegrounds") || (category == "Solo Shuffle")) {
```
We've lost this logic as far as I can tell?

tdaugaard replied:
It's no longer neccessary. See `getResultText()`, default switch case takes care of that.

aza547 commented on src/renderer/images.ts:
```diff
@@ -27,25 +29,15 @@ for (const id of arenaIDs) {
+    try {
+        dungeon[id] = require(`../../assets/dungeon/${id}.jpg`);
+    }
+    catch {};
```
probably need some logging in this branch

tdaugaard replied:
There was nothing before, as you can see, and I thought the logging would be pointless but I'd be happy to add as much `console.log` as you'd want me to :)

aza547 replied:
That'd be great - thanks!

tdaugaard replied:
Fixed in 372e3c6

tdaugaard commented:
Will add tests - thanks for pointing that out :)

aza547 commented:
Hi @tdaugaard, thanks for the PR, you've clearly put a lot of work into this. I'm really keen to get this PR into a state where I'm comfortable merging it. I think there is a bit of work to go still we're there, thoughts below. 

I've had a quick skim over as it's all I have time for tonight and made some light comments in line.

I've also run some tests on this branch and general feel is things seem pretty broken and I think you need to run the tests and do some debugging before further review. 

A few things I noticed in my testing:
- Arena test fails to get the player, spec and result. 
- Raid encounter test fails to finish and records forever. 
- I added a M+ test log and python script to main, and recording failed entirely. Pushed these tests to main, see [here](https://github.com/aza547/wow-recorder/commit/707b921dd294727e9d0605dd1e539f31db6f785a) and [here](https://github.com/aza547/wow-recorder/commit/0490bfa72ebb237d721cbca11310e6ad359d1d45).
- VideoButtons sizing have changed and looks bad.

You can find instructions on running tests [here](https://github.com/aza547/wow-recorder/blob/main/CONTRIBUTING.md#building-and-packaging).

tdaugaard commented on src/main/keystone.ts:
```diff
@@ -0,0 +1,90 @@
+    logEnd: Date
+    result?: Boolean
+
+    constructor(
```
Yeah, it's a short hand constructor that takes arguments and makes them as public members of the class as well. It's the same result as the following, in this particular case, but much more concise:

```typescript
class ChallengeModeVideoSegment {
    logEnd: Date;
    result?: Boolean;
    segmentType: VideoSegmentType;
    logStart: Date;
    ts: number;
    encounterId?: number;

    constructor(
        segmentType: VideoSegmentType,
        logStart: Date,
        ts: number,
        encounterId?: number
    ) {
        this.segmentType = segmentType;
        this.logStart = logStart;
        this.ts = ts;
        this.encounterId = encounterId;

        // Initially, let's set this to log start date to avoid logEnd
        // potentially being undefined.
        this.logEnd = logStart;
    }
};
```

aza547 replied:
neat - thanks for the education!

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -117,41 +123,173 @@ const tailFile = (path: string) => {
+ * lists, tuples, and what have we.
+ * 
+ */
+const splitLogLine = (line: string): any => {
```
The reason it looks like C code (although it's a port from a Python script I made) is that it does something low level that JS cannot do by itself.

It basically parses the log file format _accurately_ and creates arrays as needed for things like affix lists, converts the tuples from talents and such into lists.

I understand the need to keep things simple, but parsing this log file isn't simple as such. Splitting on comma without regard to encounters that contain commas is a recipe for trouble.

Example is the boss name `Skulloc, Son of Gruul` - this previously cut it short to `Skulloc`, which isn't correct.

> I'm even more nervous if this can handle the COMBATANT_DATA (which are fairly complex lines) events we rely on for getting specs/teams in arena.

This is exactly what it can do. I'd be happy to implement some tests for this, if you want.

aza547 replied:
Ah, good point. Look forward to discussing further. You might be interested in taking a look at this: https://github.com/wowarenalogs/wowarenalogs/tree/main/packages/wow-combat-log-parser/src

tdaugaard replied:
That looks very overengineered, considering it's a partial, heavily PvP centric/specific parser, at a glance.

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -117,41 +123,173 @@ const tailFile = (path: string) => {
+    let value: any = '';
+
+    // eslint-disable-next-line no-plusplus
+    // Timestamp is always a fixed width of 18, and we want to skip that here.
```
You're right, it's not true - but there's never more than 18 characters of date:

Log:

```
12/19 20:41:11.786  COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.1.5,PROJECT_ID,1
8/22 21:38:27.944  COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.2.7,PROJECT_ID,1
1/2 15:14:32.191  COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.1.5,PROJECT_ID,1
```

Input:

```js
console.log('12/19 20:41:11.786  COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.1.5,PROJECT_ID,1'.substring(18));
console.log('8/22 21:38:27.944  COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.2.7,PROJECT_ID,1'.substring(18));
console.log('1/2 15:14:32.191  COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.1.5,PROJECT_ID,1'.substring(18));
```

Output:

```
  COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.1.5,PROJECT_ID,1
 COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.2.7,PROJECT_ID,1
COMBAT_LOG_VERSION,19,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,9.1.5,PROJECT_ID,1
```

aza547 replied:
ah - good point

tdaugaard replied:
I've clarified/corrected the comment to accurately describe the situation, that is, a max length rather than a fixed length in 92821f4

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -117,41 +123,173 @@ const tailFile = (path: string) => {
+    // Get log line token
+    lineToken = lineToken.substring(0, lineToken.indexOf(','));
+
+    switch (lineToken) {
```
switch..case is always faster than doing many if..else's, especially when the value to test is a simple string and doesn't have other conditions than that.

I've no strong feelings one way or the other regarding this, besides this being faster, performance wise.

tdaugaard commented on src/main/keystone.ts:
```diff
@@ -0,0 +1,90 @@
+    }
+};
+
+const calculateCompletionResult = (mapId: number, duration: number): number => {
```
Fixed in 40848d1

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -205,35 +344,132 @@ const handleArenaStartLine = (line: string) => {
+
+    metadata = {
+        name: lineArgs[1],
+        encounterID: parseInt(lineArgs[1], 10),
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt describes that `radix` (the base) should not be assumed to be 10, and I'd rather not it be interpreted differently.

`radix` has a default, and in most cases it's 10, but it's up to the JS implementation to decide what the default is.

It's normally good practice to include it to make sure you get what you want.

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -500,17 +752,16 @@ const pollWowProcess = () => {
-    const [month, day] = date.split("/");
-    const [hours, mins, secs] = time.split(":");
-
+const getCombatLogDate = (line: string): Date => {
```
It's my habit of optimization, either for readability (for me, this is easier to read - believe it or not), or performance. I have no strong feelings one way or the other.

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -117,41 +123,173 @@ const tailFile = (path: string) => {
-    } else if (line.includes("SPELL_AURA_APPLIED")){
-        handleSpellAuraAppliedLine(line);
+    // Skip timestamp
+    let lineToken = line.substring(18);
```
See https://github.com/aza547/wow-recorder/pull/84#discussion_r964173053 for comment about this.

tdaugaard commented:
> Arena test fails to get the player, spec and result.

Fixed in b0710fa - I was parsing values starting with `0x` wrongly due to parseInt(..., 10) - needs to be parseInt(..., 16) for hexadecimal values.

tdaugaard commented:
> VideoButtons sizing have changed and looks bad.

Shouldn't have changed. Screenshot from the "Test" button test of an Arena match:

![image](https://user-images.githubusercontent.com/875107/188941418-4135154b-73d2-4edc-8b57-11ef14f12fb6.png)

aza547 commented:
> > VideoButtons sizing have changed and looks bad.
> 
> Shouldn't have changed. Screenshot from the "Test" button test of an Arena match:
> 
> ![image](https://user-images.githubusercontent.com/875107/188941418-4135154b-73d2-4edc-8b57-11ef14f12fb6.png)

But this is how it looked before:

![image](https://user-images.githubusercontent.com/32389325/188943265-04c26576-9b2c-40ca-bb22-6e6793281a97.png)

tdaugaard commented:
I had to overlay mine on yours to see the difference, but I see it now - there are some slight font size differences and the background image looks off (due to `background: contain;`). I will fix that :)

tdaugaard commented:
BTW, I _can_ fix this entire PR to _not_ include the changes to `splitLogLine`, `getCombatLogDate`, and `handleLogLine` and make a separate PR for those things.

They are not necessary for this feature, but I'd like for at least the changes to `splitLogLine` and `handleLogLine` to stay, as they do improve the performance and accuracy significantly.

Specifically, using `.split()` every time you want a value from the line is very ineffcient, as you're splitting every single time for a single value.

Using `.contains()` on multiple lines is quite a performance hit, as we have to test every `if` expression for lines we're not interested in, which is _most_ of the log lines.

I have a fix for that, too, using an array of interesting events, then doing the  `switch (line)` on only the log event type - but that's probably better suited for a different PR altogether, too.

Let me know your thoughts as well as whether you see this whole feature as I've implemented it as likely to be merged in either state as explained above.

aza547 commented:
I don't think I mind this all being in one PR and agree that lots of stuff is inefficient. Convinced by your points regarding the splitLogLine function too.  

What's the current state of the code? Have you ran the tests? Should I re-review?

tdaugaard commented:
I'm not done with the tests yet, and I still have to test the issue with the raid recording not stopping.

I'll let you know when it's ready to re-review.

One final question, I am unable to reproduce the font size difference on the VideoButton component. I can _see_ there is a difference in some things, but there are currently (after the last commit) no changes to `App.css` that isn't scoped to the `.dungeon` class, which is only applied to VideoButton's from M+ dungeons. Perhaps I'm missing a font or something since the difference is not in size, but width.

Two Arena buttons overlaid in "Difference" mode:

![image](https://user-images.githubusercontent.com/875107/188954373-0ca12cf1-7f1f-4b82-be99-7f82c97d2308.png)

aza547 commented:
> I'm not done with the tests yet, and I still have to test the issue with the raid recording not stopping.
> 
> I'll let you know when it's ready to re-review.
> 
> One final question, I am unable to reproduce the font size difference on the VideoButton component. I can _see_ there is a difference in some things, but there are currently (after the last commit) no changes of mine to `App.css` that isn't scoped to the `.dungeon` class, which is only applied to VideoButton's from M+ dungeons. Perhaps I'm missing a font or something since the difference is not in size, but width.
> 
> Two Arena buttons overlaid in "Difference" mode:
> 
> ![image](https://user-images.githubusercontent.com/875107/188954373-0ca12cf1-7f1f-4b82-be99-7f82c97d2308.png)

Strange, I'll keep this in mind when reviewing. Maybe I'll spot the cause.

aza547 commented:
Something to consider, the test script timestamps are off because it only records for a few seconds and obviously the log used in the test spans an entire dungeon duration. In other tests I've hacked the timestamps to give better test results.

You should consider that here. Obviously makes testing slightly less representative of reality but it can hit the features then rather than trying to seek to 40 min in a 10 sec vid.  

For example, see the first line [here](https://raw.githubusercontent.com/aza547/wow-recorder/main/resources/example-logs/raid.txt). 

Suspect to make the seeking work in the test you might need to carefully hack several time stamps throughout the log? Or apply some clever interpolation programmatically to them all?

tdaugaard commented:
> Suspect to make the seeking work in the test you might need to carefully hack several time stamps throughout the log? Or apply some clever interpolation programmatically to them all?

I had considered using wall clock for the time stamps instead of the log timestamps, as they might be more accurate and also would work fine with testing, but I'd have to test that first in real-time. Will probably get an opportunity to do so tomorrow.

tdaugaard commented:
Thoroughly tested this evening and it works super well, including the time stamps - they are not off by more than 500-1200 ms.
![image](https://user-images.githubusercontent.com/875107/190007635-9ea4ea45-15d1-4aea-bd7f-583f4d734a24.png)

tdaugaard commented:
Closing because it has been replaced by PR #111
